src/cts/src/TritonCTS.cpp

```diff
@@ -1652,4 +1871,222 @@ void TritonCTS::printClockNetwork(const Clock& clockNet) const
+    return;
+  }
+
+  for (TreeBuilder* registerBuilder : *builders_) {
+    if (registerBuilder->getTreeType() == registerTree) {
+      TreeBuilder* macroBuilder = registerBuilder->getParent();
+      if (macroBuilder) {
+        computeAveSinkArrivals(registerBuilder);
+        computeAveSinkArrivals(macroBuilder);
+        adjustLatencies(macroBuilder, registerBuilder);
+      }
+    }
+  }
+}
+
+void TritonCTS::computeAveSinkArrivals(TreeBuilder* builder)
+{
+  Clock clock = builder->getClock();
+  // compute average input arrival at all sinks
+  float arrival = 0.0;
+  float ins_delay = 0.0;
+  clock.forEachSink([&](const ClockInst& sink) {
+    odb::dbITerm* iterm = sink.getDbInputPin();
+    odb::dbInst* inst = iterm->getInst();
+    sta::Pin* pin = network_->dbToSta(iterm);
+    // ignore arrival fall (no inverters in current clock tree)
+    arrival
+        += openSta_->pinArrival(pin, sta::RiseFall::rise(), sta::MinMax::max());
+    // add insertion delay
+    ins_delay = 0.0;
+    sta::LibertyCell* libCell = network_->libertyCell(network_->dbToSta(inst));
+    odb::dbMTerm* mterm = iterm->getMTerm();
+    if (libCell && mterm) {
+      sta::LibertyPort* libPort
+          = libCell->findLibertyPort(mterm->getConstName());
+      if (libPort) {
+        sta::RiseFallMinMax insDelays = libPort->clockTreePathDelays();
+        if (insDelays.hasValue()) {
+          ins_delay
+              = (insDelays.value(sta::RiseFall::rise(), sta::MinMax::max())
+                 + insDelays.value(sta::RiseFall::fall(), sta::MinMax::max()))
+                / 2.0;
+        }
+      }
+    }
+    arrival += ins_delay;
+  });
+  arrival = arrival / (float) clock.getNumSinks();
+  builder->setAveSinkArrival(arrival);
+  debugPrint(
+      logger_,
+      CTS,
+      "insertion delay",
+      1,
+      "{} {}: average sink arrival is {:0.3e}",
+      (builder->getTreeType() == macroTree) ? "macro tree" : "register tree",
+      clock.getName(),
+      builder->getAveSinkArrival());
+}
+
+// Balance latencies between macro tree and register tree
+// by adding delay buffers to one tree
+void TritonCTS::adjustLatencies(TreeBuilder* macroBuilder,
+                                TreeBuilder* registerBuilder)
+{
+  // compute top buffer delays
+  computeTopBufferDelay(registerBuilder);
+  computeTopBufferDelay(macroBuilder);
+
+  float latencyDiff = macroBuilder->getAveSinkArrival()
+                      - registerBuilder->getAveSinkArrival();
+  int numBuffers = 0;
+  TreeBuilder* builder = nullptr;
+  if (latencyDiff > 0) {
+    // add buffers to register tree
+    numBuffers = (int) (latencyDiff / registerBuilder->getTopBufferDelay());
+    builder = registerBuilder;
+  } else {
+    // add buffers to macro tree (not common but why not?)
+    numBuffers
+        = (int) (std::abs(latencyDiff) / macroBuilder->getTopBufferDelay());
+    builder = macroBuilder;
+  }
+
+  numBuffers = numBuffers * options_->getDelayBufferDerate();
```

Comment:
Should this be rounded rather than truncated?

Reply:
No, we don't want to add more buffers than needed because we don't consider wire delays.  The fewer delay buffers, the better, I think.